crates/ty_python_semantic/src/site_packages.rs

```diff
@@ -530,51 +530,142 @@ impl SystemEnvironment {
+            Self::NoSitePackagesDirFound(_) => None,
+        }
+    }
+}
+
+impl std::fmt::Display for SitePackagesDiscoveryError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::CanonicalizationError(given_path, origin, _) => {
+                display_error(f, origin, given_path, "Failed to canonicalize", None)
+            }
+            Self::PathNotExecutableOrDirectory(path, origin, _) => {
+                let thing = if origin.must_point_directly_to_sys_prefix() {
+                    "directory on disk"
+                } else {
+                    "Python executable or a directory on disk"
+                };
+                display_error(
+                    f,
+                    origin,
+                    path,
+                    &format!("Invalid {origin}"),
+                    Some(&format!("does not point to a {thing}")),
+                )
+            }
+            Self::NoPyvenvCfgFile(SysPrefixPath { inner, origin }, _) => display_error(
+                f,
+                origin,
+                inner,
+                &format!("Invalid {origin}"),
+                Some("points to a broken venv with no pyvenv.cfg file"),
+            ),
+            Self::PyvenvCfgParseError(path, kind) => {
+                write!(
+                    f,
+                    "Failed to parse the `pyvenv.cfg` file at `{path}` because {kind}"
+                )
+            }
+            Self::CouldNotReadLibDirectory(SysPrefixPath { inner, origin }, _) => display_error(
+                f,
+                origin,
+                inner,
+                "Failed to iterate over the contents of the `lib` directory of the Python installation",
+                None,
+            ),
+            Self::NoSitePackagesDirFound(SysPrefixPath { inner, origin }) => display_error(
+                f,
+                origin,
+                inner,
+                &format!("Invalid {origin}"),
+                Some(
+                    "Could not find a `site-packages` directory for this Python installation/executable",
+                ),
+            ),
+        }
+    }
+}
+
+fn display_error(
+    f: &mut std::fmt::Formatter<'_>,
+    sys_prefix_origin: &SysPrefixPathOrigin,
+    given_path: &SystemPath,
+    primary_message: &str,
+    secondary_message: Option<&str>,
+) -> std::fmt::Result {
+    let fallback: &mut dyn FnMut() -> std::fmt::Result = &mut || {
+        f.write_str(primary_message)?;
+        write!(f, " `{given_path}`")?;
+        if let Some(secondary_message) = secondary_message {
+            f.write_str(": ")?;
+            f.write_str(secondary_message)?;
+        }
+        Ok(())
+    };
+
+    let SysPrefixPathOrigin::ConfigFileSetting(file, Some(range)) = sys_prefix_origin else {
+        return fallback();
+    };
+
+    let span = usize::from(range.start())..usize::from(range.end());
```

Comment:
Hmm, It definetely doesn't make me excited that we're using annotation snippet here directly.

Reply:
Agreed, but it's useful to have snippet rendering capabilities in this module for other reasons too. It might be nice in the future to render annotations if we fail to parse pyvenv.cfg files, for example. (This _would_ be a good "help wanted" task.)